File: python/cholesky.py
from __future__ import print_function
import time

from math import sqrt
from matrix_tool import zeros, print_vector, print_matrix, append_column, multiply
import gauss_klassisch as gk
from check import check_roots


def solve(A, b):
  n = len(A)
  U = getU(A)
  print_matrix(U)
  Ut = transpose(U)

  y = []
  Utb = append_column(Ut, b)
  for row in xrange(n):
    y.append(gk.backward_left(Utb, row))
  print(y)
  Uy = append_column(U, [[e] for e in y])
  x = []
  for row in xrange(n - 1, -1, -1):
    x.append(gk.backward_right(Uy, row))
  x.reverse()
  return x, U


# Ut * U = A => det(U)^2 = det(A)
def determinant(U):
  det = 1.
  for i in xrange(len(U)):
    det *= U[i][i]
  return det**2


def inverse_triangle_right(U):
  # T = U^-1
  n = len(U)
  T = zeros(n, n)
  for i in xrange(n):
    T[i][i] = 1. / U[i][i]
    for j in xrange(i + 1, n):
      for k in xrange(j):
        T[i][j] -= T[i][k] * U[k][j]
      T[i][j] /= U[j][j]
  return T


def inverse(A, U):
  n = len(U)
  T = inverse_triangle_right(U)
  print
  print("U^-1:")
  print_matrix(T)
  print("Check U^-1:")
  print_matrix(multiply(T, U))
  B = zeros(n, n)
  for i in xrange(n):
    for j in xrange(n):
      for k in xrange(min(i, j), n):
        B[i][j] += T[i][k] * T[j][k]
  return B


def getU(A):
  n = len(A)
  m = len(A[0])
  # TODO: Use util for this
  U = [[0.] * m for _ in xrange(n)]
  for i in xrange(n):
    # Plus here to preserve float
    U[i][i] += A[i][i]
    for k in xrange(i - 1, -1, -1):
      U[i][i] -= U[k][i] * U[k][i]
    U[i][i] = sqrt(U[i][i])
    for j in xrange(i + 1, m):
      # Plus here to preserve float
      U[i][j] += A[i][j]
      for k in xrange(i - 1, -1, -1):
        U[i][j] -= U[k][i] * U[k][j]
      U[i][j] /= U[i][i]
  return U


def transpose(U):
  n = len(U)
  m = len(U[0])
  Ut = zeros(m, n)
  for i in xrange(n):
    for j in xrange(m):
      Ut[j][i] = U[i][j]
  return Ut


def main():
  from constants import A2_21 as A, b2_21 as b
  start_time = time.time()
  x, U = solve(A, b)
  solve_time = time.time()
  print("Roots:")
  print_vector(x)
  check_roots(A, x, b)
  det = determinant(U)
  det_time = time.time()
  print()
  print("Determinant: {}".format(det))
  print()
  inversed = inverse(A, U)
  inverse_time = time.time()
  print("Inverse matrix:")
  print_matrix(inversed, precision=8)
  print("Check inverse matrix:")
  print_matrix(multiply(inversed, A))

  print("Time elapsed: ")
  print(
      "\tSolution:\t{} ms\n\tDeterminant:\t{} ms\n\tInversion:\t{} ms\n\tOverall:\t{} ms".
      format((solve_time - start_time) * 1000, (det_time - solve_time) * 1000, (
          inverse_time - det_time) * 1000, (inverse_time - start_time) * 1000))


if __name__ == '__main__':
  main()

File: matlab/cholesky.m
A2 = [
    [188, -32, -9, 7, -61, 49, -53, 23, -31, 62],
    [-32, 101, 11, -32, -80, 1, 53, -7, 94, -109],
    [-9, 11, 231, 96, -62, -12, 72, -40, 72, -77],
    [7, -32, 96, 174, -37, -119, 19, 7, -27, 27],
    [-61, -80, -62, -37, 184, -34, -115, 45, -53, 46],
    [49, 1, -12, -119, -34, 182, 80, -56, -4, 18],
    [-53, 53, 72, 19, -115, 80, 231, -101, 70, -50],
    [23, -7, -40, 7, 45, -56, -101, 227, -10, 23],
    [-31, 94, 72, -27, -53, -4, 70, -10, 241, -155],
    [62, -109, -77, 27, 46, 18, -50, 23, -155, 212],
];

b2 = [
    [-44],
    [74],
    [87],
    [247],
    [-39],
    [168],
    [192],
    [-199],
    [-194],
    [243],
];

function U = factorize(A)
  n = size(A, 1);
  U = zeros(n, n);
  for i = 1:n
    U(i, i) = realsqrt(A(i, i) - sum(U(1:i-1, i) .^ 2));
    for j = i+1:n
      U(i, j) = (A(i, j) - sum(U(1:i-1, i) .* U(1:i-1, j))) / U(i, i);
    end;
  end;
end;

function [B, T] = inverse(A, U)
  n = size(U, 1);
  T = zeros(n, n);
  for i = 1:n
    T(i, i) = 1 / U(i, i);
    for j = i+1:n
      T(i, j) = -sum(T(i, 1:j-1) .* U(1:j-1, j)') / U(j, j);
    end;
  end;

  B = zeros(n, n);
  for i = 1:n
    for j = 1:n
      k = min(i, j);
      B(i, j) = sum(T(i, k:n) .* T(j, k:n));
    end;
  end;
end;

tic
U = factorize(A2)
[_, y] = backward_left([U' b2])
[_, x] = backward_right([U y'])
toc
det = determinant(U) ^ 2
[B, T] = inverse(A2, U)
toc
